chapter_08/p06_towers_of_hanoi.py

```diff
@@ -1,90 +1,120 @@
-
-    def peek(self, stacknum):
-        if self.is_empty(stacknum):
-            raise Exception("Stack is empty")
-        return self.array[self.index_of_top(stacknum)]
-
-    def minimum(self, stacknum):
-        return self.minvals[self.index_of_top(stacknum)]
-
-    def is_empty(self, stacknum):
-        return self.sizes[stacknum] == 0
-
-    def is_full(self, stacknum):
-        return self.sizes[stacknum] == self.stacksize
-
-    def index_of_top(self, stacknum):
-        offset = stacknum * self.stacksize
-        return offset + self.sizes[stacknum] - 1
-
-    def size(self, stacknum):
-        return self.sizes[stacknum]
-
-
-def f(n, start, end, buff, stack):
-    if n == 1:
-        stack.push(stack.pop(start), end)
-    else:
-        f(n - 1, start, buff, end, stack)
-        f(1, start, end, buff, stack)
-        f(n - 1, buff, end, start, stack)
-
-
-def print_tower(newstack):
-    # while not newstack.is_empty(0):
-    # print(newstack.pop(0))
-    # print("".join("-" for i in range(newstack.pop(0))))
-    # while not newstack.is_empty(1):
-    # print(newstack.pop(1))
-    # print("".join("-" for i in range(newstack.pop(1))))
-    while not newstack.is_empty(2):
-        # print(newstack.pop(2))
-        print("".join("-" for i in range(newstack.pop(2))))
-
-
-def fill_tower(n):
-    newstack = MultiStack(n * 3)
-    for i in range(n, 0, -1):
-        newstack.push(i, 0)
-    return newstack
-
-
-def example():
-    n = 3
-    newstack = fill_tower(n)
-    f(n, 0, 2, 1, newstack)
-    print_tower(newstack)
+    def __init__(self, stack_size, num_stacks=3):
+        self.stack_size = stack_size
+        self.num_stacks = num_stacks
+        self.multistack = [Stack(self.stack_size) for _ in range(self.num_stacks)]
+
+    def __stack_num_check(f):
+        def wrapper(self, *args):
+            if args[0] >= self.num_stacks or args[0] < 0:
+                raise Exception("stack_num invalid")
+            return f(self, *args)
+
+        return wrapper
+
+    @__stack_num_check
+    def push(self, stack_num, val):
+        return self.multistack[stack_num].push(val)
+
+    @__stack_num_check
+    def top(self, stack_num):
+        return self.multistack[stack_num].top()
+
+    @__stack_num_check
+    def pop(self, stack_num):
+        return self.multistack[stack_num].pop()
+
+    @__stack_num_check
+    def is_empty(self, stack_num):
+        return self.multistack[stack_num].is_empty()
+
+    @__stack_num_check
+    def show_stack(self, stack_num):
+        print(str(self.multistack[stack_num]))
+
+    @__stack_num_check
+    def get_stack(self, stack_num):
+        return self.multistack[stack_num].get_stack()
+
+    def __str__(self):
+        str_result = ""
+        idx = 0
+        for stack in self.multistack:
+            str_result += f"Stack {idx}\n"
+            str_result += str(stack) + "\n"
+            idx += 1
+        return str_result
+
+    def __repr__(self):
+        return str(self)
+
+
+class TOH:
```

Comment:
In python standard practice is to use longer variable names than you might otherwise.  I'd suggest `TowerOfHanoi` here

Reply:
done